qpdf/test_parsedoffset.cc

```diff
@@ -0,0 +1,207 @@
+#include <qpdf/QPDF.hh>
+#include <qpdf/QPDFObjectHandle.hh>
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <set>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <cstdlib>
+
+void usage()
+{
+    std::cerr
+        << "Usage: test_parsedoffset INPUT.pdf"
+        << std::endl;
+}
+
+std::string make_objdesc(qpdf_offset_t offset, QPDFObjectHandle obj)
+{
+    std::stringstream ss;
+    ss << "offset = "
+       << offset
+       << " (0x"
+       << std::hex << offset << std::dec
+       << "), ";
+
+    if (obj.isIndirect())
+    {
+        ss << "indirect "
+           << obj.getObjectID()
+           << "/"
+           << obj.getGeneration()
+           << ", ";
+    }
+    else
+    {
+        ss << "direct, ";
+    }
+
+    ss << obj.getTypeName();
+
+    return ss.str();
+}
+
+void walk(size_t stream_number, QPDFObjectHandle obj,
```

Comment:
This method does not recurse into dictionaries and arrays. As far as I can see, the correctness of offsets inside nested dictionaries and arrays is not tested. Also, it doesn't look like this traverses into content streams, which also contain objects.

Reply:
> This method does not recurse into dictionaries and arrays. As far as I can see, the correctness of offsets inside nested dictionaries and arrays is not tested.

`QPDF::GetAllObjects()` gets all indirect objects.
So we only need to recurese the direct objects.
For dictionary, it worked correctly.
However, for array, it had a bug.
I've fixed it in #377.

> Also, it doesn't look like this traverses into content streams, which also contain objects.

Again, `QPDF::GetAllObjects()` gets all indirect objects.
Also the objects in content streams are got.
Thus, it traverses all objects including the objects which are in content streams.

The test file `qpdf/qtest/qpdf/digitally-signed.pdf` has some objects which are in content streams.
The output of this test `qpdf/qtest/qpdf/digitally-signed-parsedoffset.out` contains the objects which are in the content streams.